swarm/utils/metrics.go

```diff
@@ -0,0 +1,106 @@
+// Copyright 2018 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+//NOTE: The metrics package for swarm should align with the general go-ethereum metrics,
+//which uses the `rcrowley/go-metrics` package. The latter has limitations with
+//aggregating and resetting metrics, which sparked some research on the swarm team
+//regarding metrics. Therefore, this swarm metrics package introduces some indirection,
+//which abstracts a metrics API, allowing to use a different metrics library if needed.
+//If ultimately the go-metrics issues are solved,
+//it may be omitted in favor of directly using the `metrics/metrics` package in go-ethereum
+
+package utils
+
+import (
+	"github.com/ethereum/go-ethereum/metrics"
+	"net"
+	"time"
+
+	graphite "github.com/cyberdelia/go-metrics-graphite"
+	gometrics "github.com/rcrowley/go-metrics"
+)
+
+var (
+	MetricsEnabled = metrics.Enabled
+	bzzAccount     string
+	SwarmRegistry  gometrics.Registry
+)
+
+type MetricsTimer struct {
+	timer  gometrics.Timer
+	bucket string
+}
+
+func SetupMetrics(bzzAccount string) {
+	bzzAccount = bzzAccount[:12]
+	addr, _ := net.ResolveTCPAddr("tcp", "127.0.0.1:2003")
```

Comment:
I suppose that the address will be configurable in the future.

Reply:
Maybe we should do this right away

Reply:
It appears that the InfluxDB reporter works just fine, so we should probably configure it right away, rather than the Graphite reporter.

Reply:
I liked the influxdb reporter too. Just two things to consider:
* It is not supported anymore. Also this one we might need to fork
* It didn't allow to put a prefix on measurements.
The last issue is surprising, as influxdb has great support for tagging. If we fork, we may want to extend this reporter to incorporate more of the influxdb functionality

Reply:
@holisticode check my PR, I have added namespace to metrics.

The reporter is one go file, we don't really need anyone to support it for us, it works fine I think.